### ISSUE TITLE
Phase 2 Hands-On Practice: 딥러닝 기초 프로젝트 (Week 5-9)

### ISSUE THREAD

**Issue by SaraHan774** [SaraHan774/26weeks-of-ai#5]
# Phase 2: 딥러닝 기초 (Week 5-9)

> **목표**: PyTorch를 손에 익히고, 신경망이 "학습한다"는 것이 무엇인지 완전히 이해합니다.

**로드맵 연결**: `llm-learning-roadmap.md` Phase 2 (5-9주)

---

## 📋 Overview

Phase 2에서는 2개의 hands-on 프로젝트를 진행합니다:

| 주차 | 프로젝트 | 핵심 학습 | 난이도 |
|:---:|:---|:---|:---:|
| Week 5-7 | MNIST 숫자 분류기 + 학습 과정 시각화 | PyTorch, MLP, 학습 루프, Cross-Entropy | ⭐⭐ |
| Week 8-9 | Fashion-MNIST CNN 분류기 | CNN, 컨볼루션, 과적합 | ⭐⭐⭐ |

---

### 목표
손글씨 숫자 이미지를 보고 0-9를 분류하는 신경망을 PyTorch로 구현하고, 학습 과정을 시각화합니다.

### 왜 이 프로젝트인가?
- **PyTorch 입문**: 딥러닝 프레임워크의 핵심 개념 (Tensor, autograd, DataLoader)
- **학습 루프 이해**: forward → loss → backward → update 흐름 체득
- **하이퍼파라미터 실험**: 학습률, 배치 크기 변경이 미치는 영향 관찰

#### 필수 기능 (Week 5-6)
1. **데이터 로딩**
   - MNIST 데이터셋 다운로드 (torchvision)
   - DataLoader로 배치 처리
   - 데이터 정규화 (0-255 → 0-1)

2. **MLP 모델 구현**
   ```python
   Input (28x28=784) → FC(256) → ReLU → FC(128) → ReLU → FC(10) → Softmax
   ```

3. **학습 루프**
   - Forward pass
   - Cross-Entropy Loss 계산
   - Backward pass (자동 미분)
   - Optimizer step (SGD 또는 Adam)

4. **평가**
   - 테스트 정확도 95% 이상 달성
   - Confusion matrix 출력

#### 확장 기능 (Week 7)
5. **학습 과정 시각화**
   - Epoch별 Loss 그래프
   - Train/Validation 정확도 비교
   - 잘못 분류한 샘플 시각화

6. **하이퍼파라미터 실험**
   - 학습률: 0.001, 0.01, 0.1 비교
   - 배치 크기: 32, 64, 128 비교
   - Optimizer: SGD vs Adam 비교
   - 결과를 표로 정리

### 기술 스택
```python
- PyTorch
- torchvision
- matplotlib
- scikit-learn (confusion matrix)
```

#### Step 1: 환경 세팅
```bash
mkdir week05-mnist
cd week05-mnist

python -m venv venv
source venv/bin/activate

pip install torch torchvision matplotlib scikit-learn
```

#### Step 2: 코드 구조
```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import datasets, transforms
import matplotlib.pyplot as plt

# 1. 데이터 준비
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.1307,), (0.3081,))  # MNIST mean, std
])

train_dataset = datasets.MNIST('./data', train=True, download=True, transform=transform)
test_dataset = datasets.MNIST('./data', train=False, transform=transform)

train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=1000, shuffle=False)

# 2. 모델 정의
class MLP(nn.Module):
    def __init__(self):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(28*28, 256)
        self.fc2 = nn.Linear(256, 128)
        self.fc3 = nn.Linear(128, 10)
        self.relu = nn.ReLU()
    
    def forward(self, x):
        x = x.view(-1, 28*28)  # Flatten
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.fc3(x)
        return x

model = MLP()

# 3. 손실 함수 & 옵티마이저
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# 4. 학습 루프
def train_one_epoch(model, loader, criterion, optimizer):
    model.train()
    total_loss = 0
    correct = 0
    
    for batch_idx, (data, target) in enumerate(loader):
        optimizer.zero_grad()           # 그래디언트 초기화
        output = model(data)            # Forward
        loss = criterion(output, target) # Loss 계산
        loss.backward()                 # Backward
        optimizer.step()                # 파라미터 업데이트
        
        total_loss += loss.item()
        pred = output.argmax(dim=1)
        correct += pred.eq(target).sum().item()
    
    avg_loss = total_loss / len(loader)
    accuracy = 100. * correct / len(loader.dataset)
    return avg_loss, accuracy

# 5. 평가 함수
def evaluate(model, loader, criterion):
    model.eval()
    total_loss = 0
    correct = 0
    
    with torch.no_grad():  # 그래디언트 계산 안 함
        for data, target in loader:
            output = model(data)
            loss = criterion(output, target)
            total_loss += loss.item()
            pred = output.argmax(dim=1)
            correct += pred.eq(target).sum().item()
    
    avg_loss = total_loss / len(loader)
    accuracy = 100. * correct / len(loader.dataset)
    return avg_loss, accuracy

# 6. 학습 실행
epochs = 10
train_losses, train_accs = [], []
test_losses, test_accs = [], []

for epoch in range(1, epochs + 1):
    train_loss, train_acc = train_one_epoch(model, train_loader, criterion, optimizer)
    test_loss, test_acc = evaluate(model, test_loader, criterion)
    
    train_losses.append(train_loss)
    train_accs.append(train_acc)
    test_losses.append(test_loss)
    test_accs.append(test_acc)
    
    print(f'Epoch {epoch}: Train Loss={train_loss:.4f}, Train Acc={train_acc:.2f}%, '
          f'Test Loss={test_loss:.4f}, Test Acc={test_acc:.2f}%')

# 7. 시각화
plt.figure(figsize=(12, 4))
plt.subplot(1, 2, 1)
plt.plot(train_losses, label='Train Loss')
plt.plot(test_losses, label='Test Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend()

plt.subplot(1, 2, 2)
plt.plot(train_accs, label='Train Acc')
plt.plot(test_accs, label='Test Acc')
plt.xlabel('Epoch')
plt.ylabel('Accuracy (%)')
plt.legend()

plt.tight_layout()
plt.savefig('training_curves.png')
plt.show()
```

#### PyTorch 핵심 개념
1. **Tensor**: NumPy 배열과 비슷하지만 GPU 사용 가능
   ```python
   x = torch.tensor([1, 2, 3])
   x = x.cuda()  # GPU로 이동 (GPU 있을 경우)
   ```

2. **Autograd**: 자동 미분
   ```python
   x = torch.tensor([2.0], requires_grad=True)
   y = x ** 2
   y.backward()  # dy/dx 계산
   print(x.grad)  # 4.0
   ```

3. **nn.Module**: 모델 정의의 기본
   - `__init__`: 레이어 정의
   - `forward`: 순전파 로직

4. **DataLoader**: 배치 처리 자동화
   - `shuffle=True`: 매 epoch마다 섞기
   - `batch_size`: 한 번에 처리할 샘플 수

#### Cross-Entropy Loss의 의미
- 모델의 "확신도"와 "정답"의 차이를 측정
- 분류 문제에 최적화된 손실 함수
- `nn.CrossEntropyLoss`는 Softmax + NLLLoss 포함

### 완료 기준
- [ ] MNIST 데이터셋 로딩 및 시각화
- [ ] MLP 모델 정의 및 학습
- [ ] 테스트 정확도 95% 이상 달성
- [ ] 학습 곡선 그래프 저장
- [ ] 하이퍼파라미터 실험 결과 정리

---

### 목표
의류 이미지를 10개 카테고리로 분류하는 CNN을 구현하고, MLP와 성능을 비교합니다.

### 왜 이 프로젝트인가?
- **CNN 이해**: 컨볼루션, 풀링이 어떻게 작동하는지
- **Feature Map 시각화**: CNN이 무엇을 "보는지" 관찰
- **과적합 극복**: Dropout, Data Augmentation

#### 필수 기능 (Week 8)
1. **CNN 모델 구현**
   ```
   Input (28x28x1)
   → Conv2d(32, 3x3) → ReLU → MaxPool(2x2)
   → Conv2d(64, 3x3) → ReLU → MaxPool(2x2)
   → Flatten → FC(128) → ReLU → Dropout(0.5) → FC(10)
   ```

2. **학습 & 평가**
   - Fashion-MNIST 데이터셋 사용
   - 10 epochs 학습
   - 정확도 85% 이상 목표

3. **MLP vs CNN 비교**
   - 동일 데이터로 MLP와 CNN 학습
   - 정확도, 학습 속도 비교

#### 확장 기능 (Week 9)
4. **Feature Map 시각화**
   - 첫 번째 Conv 레이어의 필터 시각화
   - 각 레이어의 activation map 출력

5. **데이터 증강 (Data Augmentation)**
   ```python
   transforms.RandomRotation(10)
   transforms.RandomHorizontalFlip()
   ```

6. **잘못 분류된 샘플 분석**
   - Confusion matrix 생성
   - 가장 헷갈리는 카테고리 쌍 찾기

#### CNN 모델 구조
```python
class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.fc1 = nn.Linear(64 * 7 * 7, 128)
        self.fc2 = nn.Linear(128, 10)
        self.dropout = nn.Dropout(0.5)
        self.relu = nn.ReLU()
    
    def forward(self, x):
        # Input: (batch, 1, 28, 28)
        x = self.pool(self.relu(self.conv1(x)))  # (batch, 32, 14, 14)
        x = self.pool(self.relu(self.conv2(x)))  # (batch, 64, 7, 7)
        x = x.view(-1, 64 * 7 * 7)  # Flatten
        x = self.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

model = CNN()
print(f"Total parameters: {sum(p.numel() for p in model.parameters())}")
```

#### Feature Map 시각화
```python
def visualize_feature_maps(model, image):
    """첫 번째 Conv 레이어의 출력 시각화"""
    model.eval()
    with torch.no_grad():
        # 첫 번째 Conv 레이어만 통과
        x = model.conv1(image.unsqueeze(0))
        x = model.relu(x)
    
    # 32개 필터의 출력
    feature_maps = x.squeeze(0).cpu().numpy()
    
    fig, axes = plt.subplots(4, 8, figsize=(16, 8))
    for i, ax in enumerate(axes.flat):
        if i < len(feature_maps):
            ax.imshow(feature_maps[i], cmap='viridis')
            ax.axis('off')
    plt.tight_layout()
    plt.savefig('feature_maps.png')
    plt.show()

# 사용 예시
sample_image, _ = test_dataset[0]
visualize_feature_maps(model, sample_image)
```

#### Confusion Matrix
```python
from sklearn.metrics import confusion_matrix
import seaborn as sns

def plot_confusion_matrix(model, loader):
    model.eval()
    all_preds = []
    all_targets = []
    
    with torch.no_grad():
        for data, target in loader:
            output = model(data)
            preds = output.argmax(dim=1)
            all_preds.extend(preds.cpu().numpy())
            all_targets.extend(target.cpu().numpy())
    
    cm = confusion_matrix(all_targets, all_preds)
    
    plt.figure(figsize=(10, 8))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues')
    plt.xlabel('Predicted')
    plt.ylabel('True')
    plt.title('Confusion Matrix')
    plt.savefig('confusion_matrix.png')
    plt.show()

# Fashion-MNIST 클래스 이름
classes = ['T-shirt', 'Trouser', 'Pullover', 'Dress', 'Coat',
           'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
```

#### CNN의 장점
1. **공간 정보 보존**: Flatten하지 않고 2D 구조 유지
2. **파라미터 공유**: 같은 필터를 이미지 전체에 적용
3. **평행 이동 불변성**: 물체 위치가 바뀌어도 인식

#### 컨볼루션 연산 이해
```python
# 3x3 커널로 5x5 이미지 컨볼루션
# Output size = (Input - Kernel + 2*Padding) / Stride + 1
# (5 - 3 + 2*0) / 1 + 1 = 3

# Padding=1을 사용하면 크기 유지
# (5 - 3 + 2*1) / 1 + 1 = 5
```

#### 과적합 방지
- **Dropout**: 학습 시 랜덤하게 뉴런 비활성화
- **Data Augmentation**: 데이터를 변형하여 다양성 증가
- **Early Stopping**: Validation loss가 증가하면 중단

### 완료 기준
- [ ] CNN 모델 구현 및 학습
- [ ] 정확도 85% 이상 달성
- [ ] MLP vs CNN 성능 비교표 작성
- [ ] Feature map 시각화
- [ ] Confusion matrix 분석

---

### Week 9 복습 과제
1. **"신경망 학습 원리" 설명글 작성**
   - 주제: "컴퓨터는 어떻게 이미지를 '학습'하는가?"
   - 포함 내용:
     - Forward/Backward propagation
     - Loss function의 역할
     - Gradient descent 직관
     - CNN이 MLP보다 나은 이유
   
2. **코드 재구현**
   - 문서 없이 MNIST 학습 코드를 처음부터 다시 작성
   - 막히는 부분이 있다면 복습 필요

### 학습 체크리스트
- [ ] PyTorch Tensor 연산 이해
- [ ] Autograd 작동 원리 이해
- [ ] nn.Module 상속 및 forward 정의
- [ ] DataLoader 사용법
- [ ] 학습 루프 (train/eval 모드)
- [ ] CrossEntropyLoss 이해
- [ ] Conv2d, MaxPool2d 이해
- [ ] 과적합 vs 일반화 개념

---

### 필수 자료
1. **PyTorch 60 Minute Blitz**
   - https://pytorch.org/tutorials/beginner/deep_learning_60min_blitz.html
   - Week 5-6: PyTorch 기초

2. **밑바닥부터 시작하는 딥러닝 4-6장**
   - Week 5-7: 신경망 학습 원리

3. **3Blue1Brown - Neural Networks 시리즈**
   - https://www.youtube.com/playlist?list=PLZHQObOWTQDNU6R1_67000Dx_ZCJB-3pi
   - Week 4-7: 시각적 이해

4. **Stanford CS231n - Lecture 5 (CNNs)**
   - http://cs231n.stanford.edu/
   - Week 8-9: CNN 아키텍처

### 추가 자료
- Fast.ai Practical Deep Learning Lesson 1-2
- StatQuest "Cross Entropy"
- Andrej Karpathy "Backpropagation explained"

---

### GPU 사용 (선택)
```python
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)

# 데이터도 GPU로 이동
for data, target in train_loader:
    data, target = data.to(device), target.to(device)
```

### 모델 저장 & 로드
```python
# 저장
torch.save(model.state_dict(), 'mnist_model.pth')

# 로드
model = MLP()
model.load_state_dict(torch.load('mnist_model.pth'))
model.eval()
```

### 학습이 안 될 때 체크리스트
- [ ] 학습률이 너무 크거나 작지 않은가? (0.0001 ~ 0.01)
- [ ] Loss가 NaN이 되지 않는가?
- [ ] 데이터가 제대로 정규화되었는가?
- [ ] Optimizer가 제대로 업데이트하는가?
- [ ] 모델이 train() 모드인가?

---

**다음 단계**: Phase 3 (NLP 기초)에서 텍스트를 다루는 신경망을 배웁니다!

**질문이나 막히는 부분이 있다면 이 이슈에 코멘트로 남겨주세요!** 🚀
